File: windows/EMWaver/Services/Cloud/RemoteControlClientService.cs
using EMWaver.Scripting;
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace EMWaver.Services.Cloud;

internal sealed class RemoteControlClientService
{
    internal interface IDelegate
    {
        void OnStatus(string status);
        void OnAttached(string hostSessionId);
        void OnScriptStarted(string hostSessionId, string scriptInstanceId, string? name);
        void OnUiSnapshot(string hostSessionId, string scriptInstanceId, int rev, ScriptTree? tree);
        void OnError(string message);
    }

    private readonly CloudConfig _cfg;
    private readonly CloudAuthManager _auth;

    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;

    private string? _attachedHostId;

    internal IDelegate? Delegate { get; set; }

    internal RemoteControlClientService(CloudConfig cfg, CloudAuthManager auth)
    {
        _cfg = cfg;
        _auth = auth;
    }

    internal void Stop()
    {
        try { _cts?.Cancel(); } catch { }
        _cts = null;
        try { _ws?.Abort(); } catch { }
        _ws = null;
        _attachedHostId = null;
        Delegate?.OnStatus("disconnected");
    }

    internal void ConnectAndAttach(string hostSessionId)
    {
        Stop();
        _cts = new CancellationTokenSource();
        _ = Task.Run(() => ConnectOnceAsync(hostSessionId, _cts.Token));
    }

    internal void RunScript(string name, string source)
    {
        var hostId = _attachedHostId;
        if (string.IsNullOrWhiteSpace(hostId)) return;
        _ = SendJsonAsync(new { type = "script.run", hostSessionId = hostId, name, source }, CancellationToken.None);
    }

    internal void SendUiEvent(string scriptInstanceId, int baseRev, string targetNodeId, string eventName, object? value)
    {
        var hostId = _attachedHostId;
        if (string.IsNullOrWhiteSpace(hostId)) return;

        var payload = new Dictionary<string, object?>();
        if (value != null) payload["value"] = value;

        _ = SendJsonAsync(new
        {
            type = "ui.event",
            hostSessionId = hostId,
            scriptInstanceId,
            baseRev,
            targetNodeId,
            name = eventName,
            payload,
        }, CancellationToken.None);
    }

    private async Task ConnectOnceAsync(string hostSessionId, CancellationToken ct)
    {
        try
        {
            Delegate?.OnStatus("connecting");

            var allowAnon = (Environment.GetEnvironmentVariable("EMWAVER_ALLOW_ANON_SYNC") ?? "") == "1";
            var tok = _auth.GetIdToken();
            if (string.IsNullOrWhiteSpace(tok) && !allowAnon)
            {
                Delegate?.OnError("missing auth token");
                return;
            }

            var baseUrl = _cfg.BackendBaseUrl?.Trim() ?? "";
            if (string.IsNullOrWhiteSpace(baseUrl))
            {
                Delegate?.OnError("missing backend url");
                return;
            }

            var wsUrl = baseUrl.TrimEnd('/');
            wsUrl = wsUrl.Replace("https://", "wss://").Replace("http://", "ws://");
            wsUrl = wsUrl + "/v1/ws";
            if (!string.IsNullOrWhiteSpace(tok))
            {
                wsUrl += "?token=" + Uri.EscapeDataString(tok.Trim());
            }

            var ws = new ClientWebSocket();
            _ws = ws;
            await ws.ConnectAsync(new Uri(wsUrl), ct);
            Delegate?.OnStatus("open");

            await SendJsonAsync(new { type = "hello", role = "web", protocolVersion = 1 }, ct);
            await SendJsonAsync(new { type = "host.attach", hostSessionId }, ct);

            await ReceiveLoopAsync(ct);
        }
        catch (Exception ex)
        {
            Delegate?.OnStatus("error");
            Delegate?.OnError(ex.Message);
        }
    }

    private async Task ReceiveLoopAsync(CancellationToken ct)
    {
        var buf = new byte[256 * 1024];
        while (!ct.IsCancellationRequested && _ws != null && _ws.State == WebSocketState.Open)
        {
            WebSocketReceiveResult res;
            var ms = new System.IO.MemoryStream();
            do
            {
                res = await _ws.ReceiveAsync(new ArraySegment<byte>(buf), ct);
                if (res.MessageType == WebSocketMessageType.Close)
                {
                    Delegate?.OnStatus("closed");
                    return;
                }
                ms.Write(buf, 0, res.Count);
            } while (!res.EndOfMessage);

            var text = Encoding.UTF8.GetString(ms.ToArray());
            HandleIncoming(text);
        }
    }

    private void HandleIncoming(string text)
    {
        try
        {
            using var doc = JsonDocument.Parse(text);
            var root = doc.RootElement;
            var type = root.TryGetProperty("type", out var t) ? (t.GetString() ?? "") : "";

            if (type == "host.attached")
            {
                _attachedHostId = root.TryGetProperty("hostSessionId", out var hid) ? hid.GetString() : _attachedHostId;
                Delegate?.OnAttached(_attachedHostId ?? hostSessionId: "");
                return;
            }

            if (type == "host.error")
            {
                var err = root.TryGetProperty("error", out var e) ? (e.GetString() ?? "error") : "error";
                Delegate?.OnError("host error: " + err);
                return;
            }

            if (type == "script.started")
            {
                var hostId = root.TryGetProperty("hostSessionId", out var hh) ? (hh.GetString() ?? "") : (_attachedHostId ?? "");
                var scriptId = root.TryGetProperty("scriptInstanceId", out var si) ? (si.GetString() ?? "") : "";
                var name = root.TryGetProperty("name", out var n) ? n.GetString() : null;
                Delegate?.OnScriptStarted(hostId, scriptId, name);
                return;
            }

            if (type == "ui.snapshot")
            {
                var hostId = root.TryGetProperty("hostSessionId", out var hh) ? (hh.GetString() ?? "") : (_attachedHostId ?? "");
                var scriptId = root.TryGetProperty("scriptInstanceId", out var si) ? (si.GetString() ?? "") : "";
                var rev = root.TryGetProperty("rev", out var rr) && rr.TryGetInt32(out var rri) ? rri : 0;

                ScriptTree? tree = null;
                if (root.TryGetProperty("root", out var rootNode) && rootNode.ValueKind == JsonValueKind.Object)
                {
                    var node = DecodeNode(rootNode);
                    if (node != null)
                    {
                        tree = new ScriptTree { Root = node, Metadata = new Dictionary<string, object?>() };
                    }
                }

                Delegate?.OnUiSnapshot(hostId, scriptId, rev, tree);
                return;
            }

            if (type == "script.error")
            {
                var err = root.TryGetProperty("error", out var e) ? (e.GetString() ?? "error") : "error";
                Delegate?.OnError("script error: " + err);
                return;
            }

            if (type == "error")
            {
                var err = root.TryGetProperty("error", out var e) ? (e.GetString() ?? "error") : "error";
                Delegate?.OnError(err);
                return;
            }
        }
        catch
        {
        }
    }

    private static ScriptNode? DecodeNode(JsonElement n)
    {
        try
        {
            var id = n.TryGetProperty("id", out var i) ? (i.GetString() ?? "") : "";
            var typeRaw = n.TryGetProperty("type", out var tr) ? (tr.GetString() ?? "column") : "column";
            if (!ScriptNodeTypeExtensions.TryFromRaw(typeRaw, out var type)) type = ScriptNodeType.Column;

            var raw = new Dictionary<string, object?>();
            if (n.TryGetProperty("props", out var props) && props.ValueKind == JsonValueKind.Object)
            {
                foreach (var p in props.EnumerateObject())
                {
                    raw[p.Name] = p.Value.ToString();
                }
            }

            var handlers = new Dictionary<ScriptEventType, string>();
            if (n.TryGetProperty("handlers", out var hh) && hh.ValueKind == JsonValueKind.Object)
            {
                foreach (var h in hh.EnumerateObject())
                {
                    if (ScriptEventTypeExtensions.TryFromRaw(h.Name, out var ev))
                    {
                        handlers[ev] = h.Value.GetString() ?? "";
                    }
                }
            }

            var kids = new List<ScriptNode>();
            if (n.TryGetProperty("children", out var ch) && ch.ValueKind == JsonValueKind.Array)
            {
                foreach (var c in ch.EnumerateArray())
                {
                    var child = DecodeNode(c);
                    if (child != null) kids.Add(child);
                }
            }

            return new ScriptNode
            {
                Id = id,
                Type = type,
                Props = new ScriptNodeProps(raw, handlers),
                Children = kids,
            };
        }
        catch
        {
            return null;
        }
    }

    private async Task SendJsonAsync(object obj, CancellationToken ct)
    {
        try
        {
            if (_ws == null || _ws.State != WebSocketState.Open) return;
            var json = JsonSerializer.Serialize(obj);
            var bytes = Encoding.UTF8.GetBytes(json);
            await _ws.SendAsync(bytes, WebSocketMessageType.Text, true, ct);
        }
        catch
        {
        }
    }
}

File: macos/EMWaver/EMWaver/RemoteControlClientService.swift
import Combine
import Foundation

import EMWaverScriptModel

@MainActor
final class RemoteControlClientService: ObservableObject {
    @Published private(set) var wsStatus: String = "disconnected"
    @Published private(set) var attachedHostSessionId: String?
    @Published private(set) var remoteScriptTree: ScriptTree?
    @Published private(set) var remoteActiveScriptName: String?
    @Published private(set) var remoteScriptInstanceId: String?
    @Published private(set) var uiRev: Int = 0
    @Published var lastError: String?

    private let urlSession: URLSession

    private weak var auth: AuthenticationManager?
    private weak var hostSessions: HostSessionManager?

    private var socket: URLSessionWebSocketTask?
    private var reconnectTask: Task<Void, Never>?

    private var targetHostSessionId: String?

    init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }

    func start(auth: AuthenticationManager, hostSessions: HostSessionManager) {
        self.auth = auth
        self.hostSessions = hostSessions
    }

    func stop() {
        reconnectTask?.cancel()
        reconnectTask = nil
        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil
        wsStatus = "disconnected"
        attachedHostSessionId = nil
        targetHostSessionId = nil
    }

    func attach(to hostSessionId: String) {
        lastError = nil
        remoteScriptTree = nil
        remoteActiveScriptName = nil
        remoteScriptInstanceId = nil
        uiRev = 0
        tokenIndex = [:]

        targetHostSessionId = hostSessionId
        reconnectTask?.cancel()
        reconnectTask = Task { [weak self] in
            guard let self else { return }
            await self.reconnectLoop()
        }
    }

    func runScript(name: String, source: String) {
        guard let socket else { return }
        guard let hostId = attachedHostSessionId else { return }
        sendJson([
            "type": "script.run",
            "hostSessionId": hostId,
            "name": name,
            "source": source,
        ])
    }

    func sendUiEvent(targetNodeId: String, event: ScriptEventType, value: Any?) {
        guard let socket else { return }
        guard let hostId = attachedHostSessionId else { return }
        guard let scriptInstanceId = remoteScriptInstanceId else { return }

        var payload: [String: Any] = [:]
        if let value, (event == .change || event == .select || event == .submit) {
            payload["value"] = value
        }

        sendJson([
            "type": "ui.event",
            "hostSessionId": hostId,
            "scriptInstanceId": scriptInstanceId,
            "baseRev": uiRev,
            "targetNodeId": targetNodeId,
            "name": event.rawValue,
            "payload": payload,
        ])
    }

    // MARK: - WebSocket

    private func backendWsUrl() -> URL? {
        guard let auth else { return nil }

        let envURL = (ProcessInfo.processInfo.environment["EMWAVER_BACKEND_URL"] ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let defaultsURL = (UserDefaults.standard.string(forKey: "emwaver.agent.backendURL") ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        let raw = !envURL.isEmpty ? envURL : defaultsURL
        guard !raw.isEmpty, var base = URL(string: raw) else { return nil }
        base.appendPathComponent("v1/ws")

        let allowAnonSync = (ProcessInfo.processInfo.environment["EMWAVER_ALLOW_ANON_SYNC"] == "1")

        let tok: String
        if let session = auth.session, !session.idToken.isEmpty {
            tok = session.idToken
        } else if allowAnonSync {
            tok = ""
        } else {
            return nil
        }

        var comps = URLComponents(url: base, resolvingAgainstBaseURL: false)
        var q = comps?.queryItems ?? []
        if !tok.isEmpty {
            q.append(URLQueryItem(name: "token", value: tok))
        }
        comps?.queryItems = q

        guard let final = comps?.url else { return nil }

        var wsComps = URLComponents(url: final, resolvingAgainstBaseURL: false)
        if wsComps?.scheme == "https" {
            wsComps?.scheme = "wss"
        } else if wsComps?.scheme == "http" {
            wsComps?.scheme = "ws"
        }
        return wsComps?.url
    }

    private func reconnectLoop() async {
        while !Task.isCancelled {
            if socket == nil {
                await connectOnce()
            }
            try? await Task.sleep(nanoseconds: 2_000_000_000)
        }
    }

    private func connectOnce() async {
        guard let wsURL = backendWsUrl() else {
            lastError = "missing backend url or auth"
            return
        }
        guard let targetHostSessionId else { return }

        wsStatus = "connecting"

        let task = urlSession.webSocketTask(with: wsURL)
        socket = task
        task.resume()
        wsStatus = "open"

        // hello + attach (controller role)
        sendJson([
            "type": "hello",
            "role": "web",
            "protocolVersion": 1,
        ])
        sendJson([
            "type": "host.attach",
            "hostSessionId": targetHostSessionId,
        ])

        await receiveLoop()

        socket?.cancel(with: .goingAway, reason: nil)
        socket = nil
        wsStatus = "closed"
        attachedHostSessionId = nil
    }

    private func receiveLoop() async {
        while !Task.isCancelled, let socket {
            do {
                let msg = try await socket.receive()
                switch msg {
                case .string(let s):
                    handleIncoming(text: s)
                case .data(let d):
                    if let s = String(data: d, encoding: .utf8) {
                        handleIncoming(text: s)
                    }
                @unknown default:
                    break
                }
            } catch {
                break
            }
        }
    }

    private func handleIncoming(text: String) {
        guard let data = text.data(using: .utf8) else { return }
        guard let objAny = try? JSONSerialization.jsonObject(with: data) else { return }
        guard let obj = objAny as? [String: Any] else { return }

        let type = (obj["type"] as? String) ?? ""

        switch type {
        case "host.attached":
            let hostId = (obj["hostSessionId"] as? String) ?? targetHostSessionId
            attachedHostSessionId = hostId

        case "host.error":
            lastError = "host error: \((obj["error"] as? String) ?? "error")"

        case "script.started":
            remoteScriptInstanceId = (obj["scriptInstanceId"] as? String)
            remoteActiveScriptName = (obj["name"] as? String)
            uiRev = 0

        case "ui.snapshot":
            if let rev = obj["rev"] as? NSNumber { uiRev = rev.intValue }
            if let root = obj["root"] {
                if let node = decodeNode(root) {
                    let metadata = (obj["metadata"] as? [String: Any]) ?? [:]
                    remoteScriptTree = ScriptTree(root: node, metadata: metadata)
                }
            } else {
                remoteScriptTree = nil
            }

        case "script.error":
            lastError = "script error: \((obj["error"] as? String) ?? "error")"

        case "error":
            lastError = String(describing: obj["error"] ?? "error")

        default:
            break
        }
    }

    // MARK: - Decode

    private var tokenIndex: [String: (nodeId: String, event: ScriptEventType)] = [:]

    func invokeToken(token: String, args: [Any]) {
        guard let hostId = attachedHostSessionId else { return }
        guard let scriptInstanceId = remoteScriptInstanceId else { return }
        guard let target = tokenIndex[token] else { return }

        var value: Any? = nil
        if (target.event == .change || target.event == .select || target.event == .submit), let first = args.first {
            value = first
        }

        sendUiEvent(targetNodeId: target.nodeId, event: target.event, value: value)
    }

    private func decodeNode(_ any: Any) -> ScriptNode? {
        guard let d = any as? [String: Any] else { return nil }
        let id = (d["id"] as? String) ?? ""
        let typeRaw = (d["type"] as? String) ?? "column"
        let type = ScriptNodeType(rawValue: typeRaw) ?? .column

        let propsRaw = (d["props"] as? [String: Any]) ?? [:]
        var handlers: [ScriptEventType: String] = [:]
        if let h = d["handlers"] as? [String: Any] {
            for (k, v) in h {
                guard let ev = ScriptEventType(rawValue: k) else { continue }
                if let s = v as? String {
                    handlers[ev] = s
                    tokenIndex[s] = (nodeId: id, event: ev)
                }
            }
        }

        var children: [ScriptNode] = []
        if let kids = d["children"] as? [Any] {
            for it in kids {
                if let n = decodeNode(it) {
                    children.append(n)
                }
            }
        }

        return ScriptNode(id: id, type: type, props: ScriptNodeProps(raw: propsRaw, eventHandlers: handlers), children: children)
    }

    private func sendJson(_ obj: [String: Any]) {
        guard let socket else { return }
        guard JSONSerialization.isValidJSONObject(obj) else { return }
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return }
        guard let text = String(data: data, encoding: .utf8) else { return }
        socket.send(.string(text)) { _ in }
    }
}
